ciao-controller/command.go

```diff
@@ -73,6 +74,41 @@ func (c *controller) stopInstance(instanceID string) error {
 	return nil
 }
 
+// delete an instance, wait for the deleted event.
+func (c *controller) deleteInstanceSync(instanceID string) error {
+	wait := make(chan struct{})
+
+	i, err := c.ds.GetInstance(instanceID)
+	if err != nil {
+		return err
+	}
+
+	err = c.deleteInstance(instanceID)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		i.StateChange.L.Lock()
+		for i.StateLock.RLock(); i.State != payloads.Deleted && i.State != payloads.Hung; {
```

Comment:
I'm not sure if this loop is safe.  If i.State == payload.Deleted when the loop is first entered, i.StateLock will get locked, but we'll never enter the loop body and so it will not get unlocked.

Reply:
Also, if you go around the loop twice iStateLock will not be locked on the second iteration.

Reply:
The same comments applies to i.StateChange.  It may never get unlocked if we don't enter the loop and it may be unlocked when we call Wait